src/booster/builder.rs

```diff
@@ -0,0 +1,96 @@
+use serde_json::Value;
+
+use booster::Booster;
+use dataset::DataSet;
+use Error;
+use {InputMatrix, OutputVec};
+
+// types for training set
+#[derive(Clone)]
+pub struct TrainDataAdded(DataSet); // this should not implement default, so it can safely be used for construction
+#[derive(Default, Clone)]
+pub struct TrainDataNotAdded;
+
+// types for params
+#[derive(Clone)]
+pub struct ParamsAdded(Value); // this should not implement default, so it can safely be used for construction
+#[derive(Default, Clone)]
+pub struct ParamsNotAdded;
+
+/// Builder for the Booster.
+///
+/// Uses TypeState Pattern to make sure that Training Data is added
+/// so that Validation can be synced properly and params are present for training.
+#[derive(Default, Clone)]
+pub struct BoosterBuilder<T: Clone, P: Clone> {
+    train_data: T,
+    val_data: Vec<DataSet>,
+    params: P, // after #3 should this be a struct
+}
+
+impl<T: Clone, P: Clone> BoosterBuilder<T, P> {
+    /// Returns the Builder and a clone from it. Useful if you want to train 2 models with
+    /// only a couple differences
+    pub fn duplicate(self) -> (Self, Self) {
+        (self.clone(), self)
```

Comment:
Can you elaborate why this function returns two instances of Self and why self is the second one?

Calling this like `let (other, me) = me.duplicate();` feels a bit weird at first glance.

Reply:
Your example does look clean.

It would restrict us to 2 boosters. I don't know if comparing 3 or more boosters does make any sense.

Eventually, examples like these should be part of the docs, they help understand the API better.

Reply:
If you want more than 2 boosters, you'd probably use clone again, for example if you have a Vec of params you want to test you could do

```rust
let src_bst = Booster::builder()
  .add_train_data(dataset)
  .add_val_data(another_dataset)
  .add_val_data(also_a_dataset);
let boosters = params.map(|p| src_bst.clone().add_params(p).fit())
                     .filter_map(|booster| booster.ok());
```

`duplicate()` is a bit of a special case, but I think it's nice to have.